src/main/java/games/strategy/engine/framework/GameDataManager.java

```diff
@@ -140,84 +137,33 @@ private static GameData loadGameInSerializationFormat(
-        // used for this savegame, and if so try to run it
-        try {
-          final String newClassPath = GameRunner.findOldJar(readVersion, true);
-          // ask user if we really want to do this?
-          final String messageString = "<html>This TripleA engine is version " + ClientContext.engineVersion()
-              + " and you are trying to open a savegame made with version " + readVersion.toString()
-              + "<br>However, this TripleA cannot open any savegame made by any engine other than engines with the "
-              + "same first three version numbers as it (x_x_x_x)."
-              + "<br><br>TripleA now comes with older engines included with it, and has found the engine to run this "
-              + "savegame. This is a new feature and is in 'beta' stage."
-              + "<br>It will attempt to run a new instance of TripleA using the older engine jar file, and this "
-              + "instance will only be able to play this savegame."
-              + "<br><b>You may choose to either Close or Keep the current instance of TripleA!</b> (If hosting, you "
-              + "must close it). Please report any bugs or issues."
-              + "<br><br>Do you wish to continue?</html>";
-          final String yesClose = "Yes & Close Current";
-          final String yesOpen = "Yes & Do Not Close";
-          final String cancel = "Cancel";
-          final Object[] options = new Object[] {yesClose, yesOpen, cancel};
-          final JOptionPane pane = new JOptionPane(messageString, JOptionPane.PLAIN_MESSAGE,
-              JOptionPane.YES_NO_CANCEL_OPTION, null, options, yesClose);
-          final JDialog window = pane.createDialog(null, "Run old jar to open old Save Game?");
-          window.setVisible(true);
-          final Object buttonPressed = pane.getValue();
-          if (buttonPressed == null || buttonPressed.equals(cancel)) {
-            return null;
-          }
-          final boolean closeCurrentInstance = buttonPressed.equals(yesClose);
-          GameRunner.startGame(savegamePath, newClassPath);
-          if (closeCurrentInstance) {
-            ThreadUtil.sleep(1000);
-            System.exit(0);
-          }
-        } catch (final IOException e) {
-          if (ClientFileSystemHelper.areWeOldExtraJar()) {
-            throw new IOException("<html>Please run the default TripleA and try to open this game again. "
-                + "<br>This TripleA engine is old and kept only for backwards compatibility and can only open "
-                + "savegames created by engines with these first 3 version digits: "
-                + ClientContext.engineVersion().toStringFull("_", true) + "</html>");
-          } else {
-            throw new IOException(error);
-          }
-        }
-        return null;
+            + "\nTo download the latest version of TripleA, Please visit "
+            + UrlConstants.LATEST_GAME_DOWNLOAD_WEBSITE;
+        throw new IOException(error);
```

Comment:
@ssoloff This exception is caught in [GameSelectorModel](https://github.com/triplea-game/triplea/blob/0ff483b9d93c2a1598826094d7d2786333292448/src/main/java/games/strategy/engine/framework/startup/mc/GameSelectorModel.java#L112-L125), which also has a special case for "EngineVersionExceptions", and some bad "StackTrace formatting", for all other cases.
If the engine version is too high, we should probably throw and EngineVersionException instead of an IOException, which is the current case.
I changed the code temporarily to throw such an Exception instead, but this caused no error message to pop-up at all, which is also bad.
I didn't wanted to fiddle around with the "current code", leaving an inconsistent behaviour with your Proxy Serialization code, as this code will be the future.
The Proxy code works completely different, and I'd like to hear your opinion on how to report Errors to the user code-wise, I personally dislike returning error message strings, if there's an error, and returning null if everything is fine.
This method limits the ability to return other needed things instead, which I noticed while working on the bcrypt lobby code, so IMO Custom Exceptions with user-friendly messages are the way to go.
This includes having "expected exceptions", like version incompatibility and invalid savegame file (wrong extension etc.), where the user is responsible for the error vs. "unexpected exeptions", where the cause is unclear, where the console should pop-up along with a stacktrace to report the problem to here. (Or alternatively the Logger attempt we talked about in the Logger PR).
"Converting"  all of the methods to use this system, would introduce breaking changes when it comes to "error messages" over the network (which should really not happen, we should just send a key or code, which has a mapped value on the client, in order to have localization as possible option), but all the other stuff would be feasible.

Reply:
TL;DR: I agree with 99% of what you said. :smile:

> If the engine version is too high, we should probably throw and EngineVersionException instead of an IOException, which is the current case.

Hmm, I never saw `EngineVersionException` used in the code that deals specifically with deserializing a save game file.  It looks like `EngineVersionException` is used to indicate that the current engine is incompatible with a map based on the map metadata that specifies the minimum required engine version.  So it could probably be better named to indicate this. :smile:

Note that when using serialization for our save games, it's very difficult to detect an incompatible version without an `IOException` being thrown.  We deserialize the entire `GameData` object graph at once, so if it's incompatible, an `IOException` is thrown, and we don't have an opportunity to examine the `Version` embedded in the data to possibly throw a different kind of exception that indicates a version incompatibility.

I tried to address this in the new proxy serialization format by introducing a versioning subsystem within the stream itself, but it was removed during the design review due to being identified as a YAGNI.

> I didn't wanted to fiddle around with the "current code", leaving an inconsistent behaviour with your Proxy Serialization code, as this code will be the future.

:+1: Yes, the goal should be to maintain parity between the two systems with respect to behavior in the face of errors.  Once the new system is in place, we should feel free to refactor it to behave better.  If you look through the proxy serialization code, you'll see that I pretty much follow your recommendations above.  But when I get to a boundary with the legacy system, I have to "flatten" things out to stay compatible.

> The Proxy code works completely different, and I'd like to hear your opinion on how to report Errors to the user code-wise, I personally dislike returning error message strings, if there's an error, and returning null if everything is fine.

IMO, in code, errors should be represented as `Exception`s, not `String`s.  However, on the wire, I don't see a problem with encoding them as strings.  I suspect a lot of the code you see that uses non-`null` strings for errors and a `null` string for success is a result of making it easy to code for the network stuff.

> This includes having "expected exceptions", like version incompatibility and invalid savegame file (wrong extension etc.), where the user is responsible for the error vs. "unexpected exeptions", where the cause is unclear, where the console should pop-up along with a stacktrace to report the problem to here.

An "expected" exception that is recoverable should be modeled as a checked exception.  An "unexpected" exception is usually the result of a programmer error and should be modeled as an unchecked exception.  Chapter 9 of _Effective Java_ has a lot of good tips for how to model errors in general.

(Note that with Functional Programming techniques becoming more prevalent in Java, some new practices have to be adopted due to how errors need to be propagated through an FP call chain.  However, as our use of FP in TripleA is mostly limited to streaming collections and using monads like `Optional`, the advice in _Effective Java_ should still probably weigh pretty heavily.)

With respect to checked exceptions, I typically introduce a domain-specific exception type (e.g. `MementoImportException`).  IMO, introducing subclasses of that type is only necessary if the caller needs to make a different decision on how to recover from the exception.  Using your examples above, I would agree that having a subclass for "version incompatibility" might be useful because the caller could employ a technique like the old jar file hack to recover from that error.  However, I'm not sure a subclass for "wrong extension" would be useful because I can't see how the caller could recover from that error any way other than reporting the error to the user and forcing them to select a different file.